pkg/commands/kvstoreentry/create.go

```diff
@@ -26,20 +29,275 @@ func NewCreateCommand(parent cmd.Registerer, g *global.Data, m manifest.Data) *C
-func (c *CreateCommand) Exec(_ io.Reader, out io.Writer) error {
+func (c *CreateCommand) Exec(in io.Reader, out io.Writer) error {
+	if err := c.CheckFlags(); err != nil {
+		return err
+	}
+
+	if c.stdin {
+		return c.ProcessStdin(in, out)
+	}
+
+	if c.filePath != "" {
+		return c.ProcessFile(out)
+	}
+
+	if c.dirPath != "" {
+		return c.ProcessDir(out)
+	}
+
+	if c.Input.Key == "" || c.Input.Value == "" {
+		return fsterr.ErrInvalidKVCombo
+	}
+
 	err := c.Globals.APIClient.InsertKVStoreKey(&c.Input)
 	if err != nil {
 		c.Globals.ErrLog.Add(err)
 		return err
 	}
 
-	text.Success(out, "Inserted key %s into kv store %s", c.Input.Key, c.Input.ID)
+	text.Success(out, "Inserted key %s into KV Store %s", c.Input.Key, c.Input.ID)
+
+	return nil
+}
+
+// CheckFlags ensures only one of the three specified flags are provided.
+func (c *CreateCommand) CheckFlags() error {
+	flagCount := 0
+	if c.stdin {
+		flagCount++
+	}
+	if c.filePath != "" {
+		flagCount++
+	}
+	if c.dirPath != "" {
+		flagCount++
+	}
+	if flagCount > 1 {
+		return fsterr.ErrInvalidStdinFileDirCombo
+	}
+	return nil
+}
+
+// ProcessStdin streams STDIN to the batch API endpoint.
+func (c *CreateCommand) ProcessStdin(in io.Reader, out io.Writer) error {
+	// Determine if 'in' has data available.
+	if in == nil || text.IsTTY(in) {
+		return fsterr.ErrNoSTDINData
+	}
+	return c.CallBatchEndpoint(in, out)
+}
+
+// ProcessFile streams a JSON file content to the batch API endpoint.
+func (c *CreateCommand) ProcessFile(out io.Writer) error {
+	f, err := os.Open(c.filePath)
+	if err != nil {
+		c.Globals.ErrLog.Add(err)
+		return err
+	}
+	defer func() {
+		_ = f.Close()
+	}()
+	return c.CallBatchEndpoint(f, out)
+}
+
+// ProcessDir concurrently reads files from the given directory structure and
+// uploads each file to the set-value-for-key endpoint where the filename is the
+// key and the file content is the value.
+//
+// NOTE: Unlike ProcessStdin/ProcessFile content doesn't need to be base64.
+func (c *CreateCommand) ProcessDir(out io.Writer) error {
+	path, err := filepath.Abs(c.dirPath)
+	if err != nil {
+		return err
+	}
+
+	files, err := os.ReadDir(path)
+	if err != nil {
+		return err
+	}
+
+	spinner, err := text.NewSpinner(out)
+	if err != nil {
+		return err
+	}
+
+	err = spinner.Start()
+	if err != nil {
+		return err
+	}
+	fileLength := len(files)
+	msg := "%s %d of %d files"
+	spinner.Message(fmt.Sprintf(msg, "Processing", 0, fileLength) + "...")
+
+	base := filepath.Base(path)
+	processed := make(chan struct{}, c.dirConcurrency)
+	sem := make(chan struct{}, c.dirConcurrency)
+	done := make(chan bool)
+
+	var (
+		processingErrors []ProcessErr
+		filesProcessed   uint64
+		mu               sync.Mutex
+		wg               sync.WaitGroup
+	)
+
+	go func() {
+		for range processed {
+			filesProcessed++
+			spinner.Message(fmt.Sprintf(msg, "Processing", filesProcessed, fileLength) + "...")
+			if filesProcessed >= uint64(len(files)) {
```

Comment:
Wondering what it'd look like if instead of using this condition,  the `processed` channel were closed, causing the for loop to exit? Might be able to remove `IMPORTANT ` warning below in that case.

Reply:
Fixed this by replacing done channel with atomic increment and load. Will push shortly.